### PR TITLE
Restore time_sync module on macOS

### PR DESCRIPTION
**Pull request by ludwigschwardt**
MacOS does not have the `adjtimex` system call. As a result we have stripped out the entire `time_sync` module on non-Linux platforms.

However, packages like katsdpcontroller need the `ClockState` enum without the rest. Instead of spreading the complications, restore the `time_sync` module with a fake `adjtimex` instead.

It returns the correct time but with artificially large errors in the absence of real values (and the status is always "OK").

[This was a leftover from the beginning of the year when I wanted to test katsdpcontroller on my laptop. Nearly forgot about it - it was lurking in my working copy for months 🙂 ]

### CODE REVIEW

**Review by bmerry — CHANGES_REQUESTED** [review]
I'm not wild about this approach, because it means the sensors will assert things that are just made up, rather than reporting that the values can't be measured. The only meaningful thing it returns is the current time, and that's not even reported as a sensor value. My preference would be
- the dummy `adjtimex()` function raises `NotImplementedError`
- `TimeSyncUpdater` catches `NotImplementedError` and responds by setting all the sensors to `inactive`.

Only a couple of tests would need to be disabled on MacOS. If you also added a test that mocked in the NotImplemented implementation and checked the response, you could get test coverage of that fallback on every OS.

File: src/aiokatcp/adjtimex.py
Comment (ludwigschwardt) [reply src/aiokatcp/adjtimex.py:182]: I suspect the default `modes` member was probably already 0, and this line just silently added another superfluous `mode` member.

File: src/aiokatcp/adjtimex.py
Comment (ludwigschwardt) [reply src/aiokatcp/adjtimex.py:168]: Good point. I like the Bruce-mypy tag-team clothesline 🤼‍♂️  I wanted to generate an AI picture of the event but Dall-E failed me 😂

File: .github/workflows/test.yml
Comment (ludwigschwardt) [reply .github/workflows/test.yml:16]: Sadly it doesn't, which is why some folks recommend `codecov` instead. See e.g. lemurheavy/coveralls-public#613, which started in 2015 and is still going strong in 2023.

Comment (bmerry) [reply .github/workflows/test.yml:16]: > See e.g. https://github.com/lemurheavy/coveralls-public/issues/613, which started in 2015 and is still going strong in 2023.

That seems to be related to using different CI systems though. In theory I think it's possible to do the merge within Github Actions. On the other hand, I think my suggestion of having tests that use mocking to force use of the fallback implementation will make it possible for Linux to have a superset of MacOS coverage, in which case I don't think it's necessary to run coverage on MacOS at all.

> By the way, I just reverted it to the way it was before https://github.com/ska-sa/aiokatcp/pull/70.

Not exactly the way it was, because before #70 the tests weren't run on MacOS at all.

Comment (ludwigschwardt) [reply .github/workflows/test.yml:16]: > Not exactly the way it was...

Right, forgot about that part :-)

As I discovered after you replied (blame GitHub flakiness :-)), it looks like the merge does happen. I'll try that route first.

File: src/aiokatcp/adjtimex.py
Comment (ludwigschwardt) [reply src/aiokatcp/adjtimex.py:168]: To be fair: mypy didn't like the assignment to attributes, which apparently did not bother it before (there was only one `type: ignore` on the last line, but now they all complained).  It was actually flake8 with the issue [Redefinition of unused name from line n](https://www.flake8rules.com/rules/F811.html).

**Comment by ludwigschwardt** [comment]
Thanks! I've pushed a new version incorporating your suggestions.

I can't figure out the failing tests though. It fails on py311 on both Linux and macOS, and it seems to be unrelated to my fixes. In fact, if I rerun the tests on an older commit that previously passed, it now fails there too 🤔  It also doesn't fail on my laptop. Rerunning these tests several times doesn't help, so it's not just a random fluke.

It's complaining about a closed event loop in `TestUnclosedClient`.

**Comment by bmerry** [comment]
> Thanks! I've pushed a new version incorporating your suggestions.
> 
> I can't figure out the failing tests though. It fails on py311 on both Linux and macOS, and it seems to be unrelated to my fixes. In fact, if I rerun the tests on an older commit that previously passed, it now fails there too 🤔 It also doesn't fail on my laptop. Rerunning these tests several times doesn't help, so it's not just a random fluke.
> 
> It's complaining about a closed event loop in `TestUnclosedClient`.

The Python 3.11 failures I think are caused by https://github.com/python/cpython/pull/107650, which got backported to 3.11. We'll need to investigate to see if this is a regression in Python or something that we should be handling better.

**Comment by bmerry** [comment]
> The Python 3.11 failures I think are caused by [python/cpython#107650](https://github.com/python/cpython/pull/107650), which got backported to 3.11. We'll need to investigate to see if this is a regression in Python or something that we should be handling better.

I don't think that should hold up this PR, but we should investigate it before making a new release.

**Comment by bmerry** [comment]
I've opened NGC-1089 for the Python 3.11 issue.

**Review by bmerry — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml
Comment (ludwigschwardt) [reply .github/workflows/test.yml:16]: It does. The latest results claim that `Coverage decreased (-0.1%) to 93.406%`. If you drill down to the details, you'd see two jobs (linux + macOS) and a merged total of 93.41%:

Screenshot 2023-09-11 at 11 56 34

Each job has poor coverage on `adjtimex.py` but the combined coverage is 95.52%.

Comment (bmerry) [reply .github/workflows/test.yml:16]: Nice!

**Comment by bmerry** [comment]
Also don't forget to bump the year in the copyright notice.

File: tests/test_time_sync.py
Comment (ludwigschwardt) [reply tests/test_time_sync.py:83]: I'm not getting away here :-) Done!

**Comment by ludwigschwardt** [comment]
OK, after many force-pushes I'm ready for another round 😅

File: .gitignore
Comment (bmerry) [inline .gitignore:12]: I feel like that should go into your global `core.excludesfile` rather than here, because it's specific to your development environment rather than specific to the repository.

File: tests/test_time_sync.py
Comment (bmerry) [inline tests/test_time_sync.py:95]: Not sure that docstring is accurate. It only covers one of the two parametrisations.

Come to think of it, I'm not sure parametrisation is really necessary, since they're a partition. You could just write
```python
expected_status = Sensor.Status.NOMINAL if sys.platform == "linux" else Sensor.Status.INACTIVE
```
inside the test function.

File: tests/test_time_sync.py
Comment (bmerry) [inline tests/test_time_sync.py:128]: A lot of code duplication here. I'd suggest either:
- Parametrise `test_smoke` with a boolean that says whether to mock, and conditionally use `mocker` inside the test; or
- Factor out a helper function that does all the work of the test, and have the test functions call it.

**Review by bmerry — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (ludwigschwardt) [reply .gitignore:12]: Thanks, I did not know about that file... I merely cargo-culted some other repo's ignores (katpoint, etc). Much better idea!

File: tests/test_time_sync.py
Comment (ludwigschwardt) [reply tests/test_time_sync.py:95]: Thanks, I missed the partition. I originally had tests for linux and darwin until I merged them.

File: tests/test_time_sync.py
Comment (ludwigschwardt) [reply tests/test_time_sync.py:128]: I've gone for option 1 and lost 20 lines.

**Comment by ludwigschwardt** [comment]
Thanks for the suggestions! You can check again.

**Review by bmerry — APPROVED** [review]
LGTM. Go ahead and merge - there is a separate ticket for the Python 3.11 failures.

**Pull request merged**
